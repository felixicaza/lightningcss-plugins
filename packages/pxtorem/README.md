# ⚡ lightningcss-plugin-pxtorem

The `lightningcss-plugin-pxtorem` plugin is designed to convert pixel units to rem units in your CSS, making it easier to maintain responsive and scalable designs.

> [!NOTE]
> WIP - This plugin is still a work in progress.

## ✨ Features

- ✅ Converts pixel units to rem units.
- ✅ Helps maintain responsive and scalable designs.
- ✅ Customizable options for root value, precision, properties, and more.

## ⬇️ Installation

You can install the `lightningcss-plugin-pxtorem` plugin using npm or yarn:

```sh
npm install lightningcss-plugin-pxtorem
```

or

```sh
pnpm add lightningcss-plugin-pxtorem
```

## 🚀 Usage

To use the `lightningcss-plugin-pxtorem` you need to include it in your `vite.config.js` file:

```js
import { defineConfig } from "vite";
import { composeVisitors } from "lightningcss";
import pxtorem from "lightningcss-plugin-pxtorem";

export default defineConfig({
  css: {
    transformer: "lightningcss",
    lightningcss: {
      visitor: composeVisitors([pxtorem()]),
    },
  },
});
```

This plugin also is compatible with others tools that use Vite.

## 🤝 Contributing

If you wish to contribute to this project, you can do so by reading the [contribution guide](../../CONTRIBUTING.md).

## 🙌 Credits

This plugin was highly inspired by [postcss-pxtorem](https://github.com/cuth/postcss-pxtorem).

## 📄 License

This project is licensed under the MIT License. See the [license file](../../LICENSE) for more details.
