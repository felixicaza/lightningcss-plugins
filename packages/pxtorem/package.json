{
  "name": "lightningcss-plugin-pxtorem",
  "version": "0.0.1",
  "description": "Convert px to rem in css files with LightningCSS.",
  "keywords": [
    "lightningcss",
    "lightningcss-plugin",
    "css",
    "plugin",
    "pxtorem",
    "px",
    "rem"
  ],
  "homepage": "https://github.com/felixicaza/lightningcss-plugins/packages/pxtorem#readme",
  "bugs": {
    "url": "https://github.com/felixicaza/lightningcss-plugins/issues"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/felixicaza/lightningcss-plugins.git"
  },
  "license": "MIT",
  "author": "Felix Icaza <fx.joliett17@gmail.com>",
  "sideEffects": false,
  "type": "module",
  "exports": {
    ".": "./dist/index.js"
  },
  "main": "./dist/index.js",
  "module": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsup",
    "dev": "tsup --watch",
    "prepublishOnly": "pnpm build",
    "start": "tsx src/index.ts"
  },
  "packageManager": "pnpm@10.5.2"
}
