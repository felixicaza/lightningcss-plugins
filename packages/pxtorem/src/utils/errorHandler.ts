export function validateIsNumber(value: number, name: string) {
  if (typeof value !== 'number' || isNaN(value)) {
    throw new Error(`Invalid ${name}: must be a valid number.`)
  }
}

export function validatePositiveInteger(value: number, name: string) {
  if (value < 0) {
    throw new Error(`Invalid ${name}: must not be negative.`)
  }

  if (!Number.isInteger(value)) {
    throw new Error(`Invalid ${name}: must not be a decimal.`)
  }
}
