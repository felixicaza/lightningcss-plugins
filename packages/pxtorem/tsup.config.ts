import { defineConfig } from 'tsup'

export default defineConfig([
  {
    entry: ['src/index.ts'],
    format: ['esm'],
    dts: true,
    minify: true,
    target: 'esnext'
  }
])
