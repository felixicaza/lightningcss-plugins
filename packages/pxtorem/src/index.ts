import { type LengthValue, type LengthUnit } from 'lightningcss'
import { validatePositiveInteger } from './utils/errorHandler'
import type { Config } from './contracts/config.type'

const defultConfig: Config = {
  rootValue: 16,
  unitPrecision: 4
}

/**
 * Convert pixel units to rem units.
 * @param {Object} config - The configuration object.
 * @param {number} [config.rootValue=16] - The root value for conversion. (default is 16)
 * @param {number} [config.unitPrecision=4] - The decimal precision for the converted value. (default is 4)
 * @returns {Function} `Length` function to convert pixel values to rem values used by LightningCSS.
 */
function pxtorem(config: Partial<Config> = {}) {
  const { rootValue, unitPrecision } = { ...defultConfig, ...config }

  validatePositiveInteger(rootValue, 'rootValue')
  validatePositiveInteger(unitPrecision, 'unitPrecision')

  const toFixed = (value: number, precision: number): number => {
    const factor = Math.pow(10, precision)

    return Math.round(value * factor) / factor
  }

  return {
    Length({ unit, value }: LengthValue): { unit: LengthUnit, value: number } | undefined {
      if (unit === 'px') {
        return { unit: 'rem', value: toFixed(value / rootValue, unitPrecision) }
      }
    }
  }
}

export default pxtorem
