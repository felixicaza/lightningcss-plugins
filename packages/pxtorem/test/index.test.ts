import { describe, expect, it } from 'vitest'
import { composeVisitors, transform } from 'lightningcss'
import pxtorem from '../src/index'
import { validatePositiveInteger } from '../src/utils/errorHandler'

describe('pxtorem plugin', () => {
  it('should convert pixel to rem using default configuration', () => {
    const css = 'div { margin: 16px; }'

    const output = transform({
      filename: 'input.css',
      code: Buffer.from(css),
      minify: false,
      sourceMap: false,
      visitor: composeVisitors([pxtorem()])
    }).code.toString()

    expect(output).toMatchSnapshot()
  })

  it('should not convert other units', () => {
    const css = 'div { margin: 1em; }'

    const output = transform({
      filename: 'input.css',
      code: Buffer.from(css),
      minify: false,
      sourceMap: false,
      visitor: composeVisitors([pxtorem()])
    }).code.toString()

    expect(output).toMatchSnapshot()
  })

  it('should convert pixel to rem in media queries', () => {
    const css = '@media (min-width: 768px) { div { margin: 16px; } }'

    const output = transform({
      filename: 'input.css',
      code: Buffer.from(css),
      minify: false,
      sourceMap: false,
      visitor: composeVisitors([pxtorem()])
    }).code.toString()

    expect(output).toMatchSnapshot()
  })

  it('should convert pixel to rem in keyframes', () => {
    const css = '@keyframes fadeIn { from { margin: 16px; } to { margin: 32px; } }'

    const output = transform({
      filename: 'input.css',
      code: Buffer.from(css),
      minify: false,
      sourceMap: false,
      visitor: composeVisitors([pxtorem()])
    }).code.toString()

    expect(output).toMatchSnapshot()
  })

  it('should convert pixel to rem in custom properties', () => {
    const css = ':root { --margin: 16px; } div { margin: var(--margin); }'

    const output = transform({
      filename: 'input.css',
      code: Buffer.from(css),
      minify: false,
      sourceMap: false,
      visitor: composeVisitors([pxtorem()])
    }).code.toString()

    expect(output).toMatchSnapshot()
  })

  it('should handle multiple pixel values', () => {
    const css = 'div { margin: 16px 32px; }'

    const output = transform({
      filename: 'input.css',
      code: Buffer.from(css),
      minify: false,
      sourceMap: false,
      visitor: composeVisitors([pxtorem()])
    }).code.toString()

    expect(output).toMatchSnapshot()
  })

  it('should handle nested selectors', () => {
    const css = 'div { & > span { margin: 16px; } }'

    const output = transform({
      filename: 'input.css',
      code: Buffer.from(css),
      minify: false,
      sourceMap: false,
      visitor: composeVisitors([pxtorem()])
    }).code.toString()

    expect(output).toMatchSnapshot()
  })

  it('should handle float pixel units', () => {
    const css = 'div { margin: 43.3398589px; }'

    const output = transform({
      filename: 'input.css',
      code: Buffer.from(css),
      minify: false,
      sourceMap: false,
      visitor: composeVisitors([pxtorem()])
    }).code.toString()

    expect(output).toMatchSnapshot()
  })

  it('should handle custom configuration', () => {
    const css = 'div { margin: 16px; } .foo { padding: 8px; }'

    const output = transform({
      filename: 'input.css',
      code: Buffer.from(css),
      minify: false,
      sourceMap: false,
      visitor: composeVisitors([pxtorem({ rootValue: 8, unitPrecision: 2, minValue: 10 })])
    }).code.toString()

    expect(output).toMatchSnapshot()
  })

  it('should handle custom configuration with unitPrecision', () => {
    const css = 'div { margin: 43.3398589px; }'

    const output = transform({
      filename: 'input.css',
      code: Buffer.from(css),
      minify: false,
      sourceMap: false,
      visitor: composeVisitors([pxtorem({ unitPrecision: 2 })])
    }).code.toString()

    expect(output).toMatchSnapshot()
  })

  it('should use custom rootValue', () => {
    const css = 'div { margin: 32px; }'

    const output = transform({
      filename: 'input.css',
      code: Buffer.from(css),
      minify: false,
      sourceMap: false,
      visitor: composeVisitors([pxtorem({ rootValue: 32 })])
    }).code.toString()

    expect(output).toMatchSnapshot()
  })

  it('should handle extreme pixel values', () => {
    const css = 'div { margin: 10000px; }'

    const output = transform({
      filename: 'input.css',
      code: Buffer.from(css),
      minify: false,
      sourceMap: false,
      visitor: composeVisitors([pxtorem()])
    }).code.toString()

    expect(output).toMatchSnapshot()
  })

  it('should handle very small pixel values', () => {
    const css = 'div { margin: 0.001px; }'

    const output = transform({
      filename: 'input.css',
      code: Buffer.from(css),
      minify: false,
      sourceMap: false,
      visitor: composeVisitors([pxtorem()])
    }).code.toString()

    expect(output).toMatchSnapshot()
  })

  it('should handle minimum positive value', () => {
    const css = 'div { margin: 30px; } .foo { padding: 8px; }'

    const output = transform({
      filename: 'input.css',
      code: Buffer.from(css),
      minify: false,
      sourceMap: false,
      visitor: composeVisitors([pxtorem({ minValue: 10 })])
    }).code.toString()

    expect(output).toMatchSnapshot()
  })

  it('should handle minimum negative value', () => {
    const css = 'div { margin: -5px; } .foo { top: -20px; }'

    const output = transform({
      filename: 'input.css',
      code: Buffer.from(css),
      minify: false,
      sourceMap: false,
      visitor: composeVisitors([pxtorem({ minValue: -10 })])
    }).code.toString()

    expect(output).toMatchSnapshot()
  })

  it('should handle minimum float value', () => {
    const css = 'div { margin: 0.55px; } .foo { top: 0.25px; }'

    const output = transform({
      filename: 'input.css',
      code: Buffer.from(css),
      minify: false,
      sourceMap: false,
      visitor: composeVisitors([pxtorem({ minValue: 0.5 })])
    }).code.toString()

    expect(output).toMatchSnapshot()
  })

  it('should handle minimum zero value', () => {
    const css = 'div { margin: 0px; } .foo { padding: 10px; } .bar { left: 0.25px; } .baz { top: -0.55px; }'

    const output = transform({
      filename: 'input.css',
      code: Buffer.from(css),
      minify: false,
      sourceMap: false,
      visitor: composeVisitors([pxtorem({ minValue: 0 })])
    }).code.toString()

    expect(output).toMatchSnapshot()
  })

  it('should throw error for negative rootValue', () => {
    expect(() => pxtorem({ rootValue: -1 })).toThrowError('Invalid rootValue: must not be negative.')
  })

  it('should throw error for negative unitPrecision', () => {
    expect(() => pxtorem({ unitPrecision: -1 })).toThrowError('Invalid unitPrecision: must not be negative.')
  })
})

describe('validatePositiveInteger', () => {
  it('should throw error for undefined value', () => {
    expect(() => validatePositiveInteger(undefined, 'rootValue')).toThrowError('Invalid rootValue: must be a valid number.')
  })

  it('should throw error for NaN value', () => {
    expect(() => validatePositiveInteger(NaN, 'rootValue')).toThrowError('Invalid rootValue: must be a valid number.')
  })

  it('should throw error for negative value', () => {
    expect(() => validatePositiveInteger(-1, 'rootValue')).toThrowError('Invalid rootValue: must not be negative.')
  })

  it('should throw error for decimal value', () => {
    expect(() => validatePositiveInteger(1.5, 'rootValue')).toThrowError('Invalid rootValue: must not be a decimal.')
  })

  it('should pass for valid integer value', () => {
    expect(() => validatePositiveInteger(1, 'rootValue')).not.toThrow()
  })
})
