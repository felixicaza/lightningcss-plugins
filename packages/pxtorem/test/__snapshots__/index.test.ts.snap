// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`pxtorem plugin > should convert pixel to rem in custom properties 1`] = `
":root {
  --margin: 1rem;
}

div {
  margin: var(--margin);
}
"
`;

exports[`pxtorem plugin > should convert pixel to rem in keyframes 1`] = `
"@keyframes fadeIn {
  from {
    margin: 1rem;
  }

  to {
    margin: 2rem;
  }
}
"
`;

exports[`pxtorem plugin > should convert pixel to rem in media queries 1`] = `
"@media (width >= 48rem) {
  div {
    margin: 1rem;
  }
}
"
`;

exports[`pxtorem plugin > should convert pixel to rem using default configuration 1`] = `
"div {
  margin: 1rem;
}
"
`;

exports[`pxtorem plugin > should handle custom configuration 1`] = `
"div {
  margin: 2rem;
}

.foo {
  padding: 8px;
}
"
`;

exports[`pxtorem plugin > should handle custom configuration with unitPrecision 1`] = `
"div {
  margin: 2.71rem;
}
"
`;

exports[`pxtorem plugin > should handle extreme pixel values 1`] = `
"div {
  margin: 625rem;
}
"
`;

exports[`pxtorem plugin > should handle float pixel units 1`] = `
"div {
  margin: 2.7087rem;
}
"
`;

exports[`pxtorem plugin > should handle minimum float value 1`] = `
"div {
  margin: .0344rem;
}

.foo {
  top: .25px;
}
"
`;

exports[`pxtorem plugin > should handle minimum negative value 1`] = `
"div {
  margin: -.3125rem;
}

.foo {
  top: -20px;
}
"
`;

exports[`pxtorem plugin > should handle minimum positive value 1`] = `
"div {
  margin: 1.875rem;
}

.foo {
  padding: 8px;
}
"
`;

exports[`pxtorem plugin > should handle minimum zero value 1`] = `
"div {
  margin: 0;
}

.foo {
  padding: .625rem;
}

.bar {
  left: .0156rem;
}

.baz {
  top: -.55px;
}
"
`;

exports[`pxtorem plugin > should handle multiple pixel values 1`] = `
"div {
  margin: 1rem 2rem;
}
"
`;

exports[`pxtorem plugin > should handle nested selectors 1`] = `
"div {
  & > span {
    margin: 1rem;
  }
}
"
`;

exports[`pxtorem plugin > should handle very small pixel values 1`] = `
"div {
  margin: .0001rem;
}
"
`;

exports[`pxtorem plugin > should not convert other units 1`] = `
"div {
  margin: 1em;
}
"
`;

exports[`pxtorem plugin > should use custom rootValue 1`] = `
"div {
  margin: 1rem;
}
"
`;
